validator/client/src/main/java/tech/pegasys/teku/validator/client/RuntimeProposerConfig.java

```diff
@@ -68,20 +72,42 @@ public Optional<Eth1Address> getEth1AddressForPubKey(final BLSPublicKey pubKey)
     if (config == null) {
       return Optional.empty();
     }
-    return Optional.of(config.getFeeRecipient());
+    return config.getFeeRecipient();
   }
 
-  public synchronized void addOrUpdate(
+  public Optional<UInt64> getGasLimitForPubKey(final BLSPublicKey publicKey) {
+    final Optional<Config> maybeConfig = Optional.ofNullable(proposerConfigMap.get(publicKey));
+    return maybeConfig.flatMap(Config::getGasLimit);
+  }
+
+  public synchronized void addOrUpdateFeeRecipient(
       final BLSPublicKey publicKey, final Eth1Address eth1Address) {
-    proposerConfigMap.put(publicKey, new Config(eth1Address));
+    proposerConfigMap.put(
+        publicKey, new Config(Optional.ofNullable(eth1Address), getGasLimitForPubKey(publicKey)));
     storagePath.ifPresent(this::save);
   }
 
-  public void delete(final BLSPublicKey publicKey) {
-    proposerConfigMap.remove(publicKey);
+  public synchronized void addOrUpdateGasLimit(
+      final BLSPublicKey publicKey, final UInt64 gasLimit) {
+    proposerConfigMap.put(
+        publicKey, new Config(getEth1AddressForPubKey(publicKey), Optional.ofNullable(gasLimit)));
     storagePath.ifPresent(this::save);
   }
 
+  public void deleteFeeRecipient(final BLSPublicKey publicKey) {
```

Comment:
given add and read are all synchronized, should probably make delete also synchronized..